### PR TITLE
[shard-map] add rewrite for efficient transposition

### PR DESCRIPTION
**Pull request by mattjj**
See the included JEP for details, but basically we want to fix this:

```python
import jax
import jax.numpy as jnp
from jax.sharding import Mesh, PartitionSpec as P
from jax.experimental.shard_map import shard_map

# cursed identity function
f = shard_map(lambda x: x, mesh=Mesh(jax.devices(), ('x',)),
              in_specs=P(), out_specs=P())

# transpose shorthand
t = lambda f: lambda x: \
    jax.linear_transpose(f, jax.ShapeDtypeStruct((), jnp.dtype('float32')))(x)[0]

print(jax.make_jaxpr(f)(1.))
# { lambda ; a:f32[]. let
#     b:f32[] = shard_map[
#       jaxpr={ lambda ; c:f32[]. let  in (c,) }
#       in_names=({},)
#       out_names=({},)
#     ] a
#   in (b,) }

# Transpose once...
print(jax.make_jaxpr(t(f))(1.))
# { lambda ; a:f32[]. let
#     b:f32[] = div a 8.0
#     c:f32[] = shard_map[
#       jaxpr={ lambda ; d:f32[]. let
#           e:f32[] = psum[axes=('x',) axis_index_groups=None] d
#         in (e,) }
#       in_names=({},)
#       out_names=({},)
#     ] b
#   in (c,) }

# Transpose twice...
print(jax.make_jaxpr(t(t(f)))(1.))
# { lambda ; a:f32[]. let
#     b:f32[] = div a 8.0
#     c:f32[] = shard_map[
#       jaxpr={ lambda ; d:f32[]. let
#           e:f32[] = psum[axes=('x',) axis_index_groups=None] d
#           f:f32[] = psum[axes=('x',) axis_index_groups=None] e
#         in (f,) }
#       in_names=({},)
#       out_names=({},)
#     ] b
#     g:f32[] = div c 8.0
#   in (g,) }
```

(Lightly edited jaxprs.)

Holy expression swell! As we keep transposing, we keep adding `psum`s and `div`s!

And it's not just `shard_map`; any time we have a JAX map with unmapped outputs we'd run into this problem (or not have correct transposes at all). This problem's been with us from the beginning.

This kind of transposition is inefficient, but not incorrect. That is, we get the numbers right, but we're introducing defensive communication and computation that might be redundant. Indeed, in the above example there's no need to `psum` anything, as we know the values coming in are the same across all shmapped function instances; the `psum`s are just scaling those up by a factor of `len(jax.devices())`, and the `div`s are there to compensate for that.

But our transpose logic doesn't know that: we don't track when values are invariant along a shmapped axis, hence the need for defensiveness.

This PR adds a way of tracking that invariance information. Long-run we think we want to track it in jaxpr types (aka avals), and to do this rewriting in the `jnp`/`lax` traceable layer, but to keep things self-contained inside the shmap implementation, this PR takes a more "static analysis" approach, where we have a rewrite pass which tracks this extra information and exploits it to avoid these inefficiencies.

See the JEP for details!

### CODE REVIEW

**Review by froystig — APPROVED** [review]
Well done!

**Comment by mattjj** [comment]
Something about our Google source sync system got messed up, making a new PR to merge this...
